Brooks/LanguageRenderer/NewCSharpRenderer.hs

```diff
@@ -0,0 +1,561 @@
+{-# LANGUAGE TypeFamilies #-}
+
+module LanguageRenderer.NewCSharpRenderer (
+    -- * C# Code Configuration -- defines syntax of all C# code
+    CSharpCode(..)
+) where
+
+import New (Label,
+    PackageSym(..), RenderSym(..), KeywordSym(..), PermanenceSym(..),
+    BodySym(..), BlockSym(..), ControlBlockSym(..), StateTypeSym(..),
+    UnaryOpSym(..), BinaryOpSym(..), ValueSym(..), NumericExpression(..), 
+    BooleanExpression(..), ValueExpression(..), Selector(..), FunctionSym(..), 
+    SelectorFunction(..), StatementSym(..), ControlStatementSym(..), ScopeSym(..),
+    MethodTypeSym(..), ParameterSym(..), MethodSym(..), StateVarSym(..), 
+    ClassSym(..), ModuleSym(..))
+import NewLanguageRenderer (Terminator(..), fileDoc', 
+    moduleDocD, classDocD, enumDocD,
+    enumElementsDocD, multiStateDocD, blockDocD, bodyDocD, outDocD,
+    printFileDocD, boolTypeDocD, 
+    intTypeDocD, charTypeDocD, stringTypeDocD, typeDocD, listTypeDocD, voidDocD,
+    constructDocD, stateParamDocD, paramListDocD, methodDocD, methodListDocD, 
+    stateVarDocD, stateVarListDocD, ifCondDocD, switchDocD, forDocD, 
+    forEachDocD, whileDocD, stratDocD, assignDocD, plusEqualsDocD, plusPlusDocD,
+    varDecDocD, varDecDefDocD, listDecDocD, listDecDefDocD, objDecDefDocD, 
+    constDecDefDocD, statementDocD, returnDocD,
+    commentDocD, notOpDocD, negateOpDocD, unOpDocD, equalOpDocD, 
+    notEqualOpDocD, greaterOpDocD, greaterEqualOpDocD, lessOpDocD, 
+    lessEqualOpDocD, plusOpDocD, minusOpDocD, multOpDocD, divideOpDocD, 
+    moduloOpDocD, andOpDocD, orOpDocD, binOpDocD, binOpDocD', litTrueD, litFalseD,
+    litCharD, litFloatD, litIntD, litStringD, defaultCharD, defaultFloatD, defaultIntD, 
+    defaultStringD, varDocD, extVarDocD, selfDocD, argDocD, enumElemDocD, objVarDocD, 
+    inlineIfDocD, funcAppDocD, extFuncAppDocD, stateObjDocD, listStateObjDocD, 
+    notNullDocD, listIndexExistsDocD, funcDocD, castDocD, listSetDocD, 
+    listAccessDocD, objAccessDocD, castObjDocD, breakDocD, continueDocD, 
+    staticDocD, dynamicDocD, privateDocD, publicDocD, dot, new, 
+    observerListName, doubleSlash, addCommentsDocD, callFuncParamList, 
+    getterName, setterName, setMain, setEmpty, statementsToStateVars)
+import Helpers (oneTab, tripFst, tripSnd, tripThird, liftA4, liftA5, liftA6, 
+  liftA7, liftList, lift1List, lift3Pair, lift4Pair, liftPairFst, liftTripFst)
+
+import Prelude hiding (break,print,(<>),sin,cos,tan,floor)
+import qualified Data.Map as Map (fromList,lookup)
+import Control.Applicative (Applicative, liftA2, liftA3)
+import Text.PrettyPrint.HughesPJ (Doc, text, (<>), (<+>), parens, comma, empty,
+  equals, semi, vcat, lbrace, rbrace, render, colon, render)
+
+newtype CSharpCode a = CSC {unCSC :: a}
```

Comment:
I wonder if you could `derive` `Functor` and friends?

Reply:
`Functor` can be derived if we use the DeriveFunctor language extension. When I attempted to derive `Applicative` and `Monad`, GHC gave an error and recommended I use the GeneralizedNewtypeDeriving extension. I tried using that extension but it did not help, so I think we'll need to keep the `Applicative` and `Monad` instances explicit.

Is the cost of using the DeriveFunctor extension worth it to derive `Functor`?

Reply:
@JacquesCarette here was our previous discussion of deriving Functor, etc.

Reply:
I'm pretty sure we could use the (even newer) `DerivingVia` to do it.  Maybe insert that as a comment, and we can get to it later.